src/fabric/annealing.rs

```diff
@@ -94,56 +172,134 @@ impl MeasurePair {
 
 struct PairGenerator {
     joints: Vec<JointIncident>,
-    pairs: HashMap<(usize, usize), MeasurePair>,
+    intervals: HashMap<(usize, usize), Interval>,
+    pairs: HashMap<(usize, usize), Pair>,
 }
 
 impl PairGenerator {
-    fn new(joints: Vec<JointIncident>) -> Self {
+    fn new(joints: Vec<JointIncident>, intervals: HashMap<(usize, usize), Interval>) -> Self {
         Self {
             joints,
+            intervals,
             pairs: HashMap::new(),
         }
     }
 
-    fn generate_pairs(mut self) -> impl Iterator<Item=MeasurePair> {
-        for joint in self.joints.clone() {
-            self.add_pairs_for(joint)
+    fn proximity_measures(mut self) -> impl Iterator<Item=Pair> {
+        for joint in 0..self.joints.len() {
+            self.push_proximity(joint);
         }
         self.pairs.into_values()
     }
 
-    fn add_pairs_for(&mut self, joint: JointIncident) {
-        let Some(push) = &joint.push else {
+    fn push_proximity(&mut self, joint_index: usize) {
+        let Some(push) = &self.joints[joint_index].push else {
             return;
         };
         let length_limit = push.ideal();
-        let two_steps: HashSet<_> = joint.adjacent_joints
+        let two_steps: HashSet<_> = self.joints[joint_index].adjacent_joints
             .iter()
             .flat_map(|&adjacent| self.joints[adjacent].adjacent_joints.iter())
             .collect();
         let new_pairs = self.joints
             .iter()
             .filter_map(|other_joint| {
-                if joint.index == other_joint.index {
+                if joint_index == other_joint.index {
                     return None;
                 }
-                if joint.adjacent_joints.contains(&other_joint.index) {
+                if self.joints[joint_index].adjacent_joints.contains(&other_joint.index) {
                     return None;
                 }
                 if two_steps.contains(&other_joint.index) {
                     return None;
                 }
-                let length = joint.location.distance(other_joint.location);
+                let length = self.joints[joint_index].location.distance(other_joint.location);
                 if length > length_limit {
                     return None;
                 }
-                Some(MeasurePair {
-                    alpha_index: joint.index,
+                Some(Pair {
+                    alpha_index: joint_index,
                     omega_index: other_joint.index,
-                    length,
+                    link: Link::Measure { length },
                 })
             })
             .map(|pair| (pair.key(), pair));
         self.pairs.extend(new_pairs);
     }
+
+    fn bow_tie_pulls(mut self) -> impl Iterator<Item=Pair> {
+        for interval in self.intervals.values() {
+            if interval.role != Role::Push {
+                continue;
+            }
+            let mut meeting_pairs: Vec<(Path, Path)> = vec![];
```

Comment:
```suggestion
            let mut meeting_pairs = vec![];
```
Inferred from the .push(), though you could keep it for reasons of clarity.